### ISSUE TITLE
c-ares resolver crashes when authority is set to non-existent DNS server

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#15133]
Repro:
set GRPC_DNS_RESOLVER=ares
Create channel with target   `dns://127.0.0.1:12345/some.domain.name.to.resolve.xyz`
start a call

This results in a crash here (because `lb_addrs` is nullptr).
```
#4  0x00007f2962768a24 in grpc_cares_wrapper_address_sorting_sort (lb_addrs=0x0)
    at src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc:125
125	      sizeof(address_sorting_sortable) * lb_addrs->num_addresses);
```

It basically looks like that address sorting is attempted even when grpc_ares_request results in 
`*(r->lb_addrs_out)` being null.
https://github.com/grpc/grpc/blob/2abcac2d7a1e34d562efeeecfc43f7553befd9c7/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L156

We should also have a test coverage for this case.

**Comment by jtattermusch** [comment]
The rootcause:

in `on_hostbyname_done_cb`,   the   `*lb_addresses = grpc_lb_addresses_create(0, nullptr);` is only done if `status == ARES_SUCCESS`.  If there's an error,    the list remains set to `null`.
https://github.com/grpc/grpc/blob/2abcac2d7a1e34d562efeeecfc43f7553befd9c7/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L194

**Comment by apolcyn** [comment]
@jtattermusch FTR, what I'm seeing is: not only does the channel need to be created, but the DNS request resolution also needs to be <i>cancelled</i> by the grpc channel somehow. E.g., it seems that just running the C++ interop client with a bad authority just hangs. But if e.g. a channel is closed while in the middle of resolution, <i>then</i> it seems to crash. Not sure if that's the same thing that you noticed?

**Comment by jtattermusch** [comment]
@apolcyn what you're describing might be a different issue.  This issue is basically that if the resolution returns a null list of addresses, sorting them is attempted and that crashes.

**Comment by apolcyn** [comment]
Makes sense. In that case, I'm adding a missing test for a resolution failure as a [part of 15179](https://github.com/grpc/grpc/pull/15179/files#diff-5ed2be7bbbfaf874aa606879ecaa242dR289) that repro's/covers.
